src/GraphQL/Server.hs

```diff
@@ -180,27 +196,37 @@ instance forall v. ReadValue v => ReadValue (Maybe v) where
 -- the name matches the query. Note that the name is *not* in monad m,
 -- but the value is. This is necessary so we can skip execution if the
 -- name doesn't match.
-data NamedFieldExecutor m = NamedFieldExecutor AST.Name (m GValue.Value)
-
-executeNamedField :: Monad m => AST.Field -> NamedFieldExecutor m -> m (Maybe GValue.ObjectField)
-executeNamedField (AST.Field alias name _ _ _) (NamedFieldExecutor k mValue)
-  | name == k = do
-      value <- mValue
-      let name' = fromMaybe name alias
-      pure (Just (GValue.ObjectField name' value))
-  | otherwise = pure Nothing
-
--- Deconstruct object type signature and handler value at the same
--- time and run type-directed code for each field.
-resolveField :: forall a (m :: Type -> Type). BuildFieldResolver m a => m GValue.ObjectField -> FieldHandler m a -> AST.Selection -> m GValue.ObjectField
-resolveField fallback lh selection@(AST.SelectionField field) = do
-  let fieldExecutor = buildFieldResolver @m @a lh selection
-  case fieldExecutor of
-    Left err -> throwM err
-    Right executor -> do
-      objectField <- executeNamedField field executor
-      maybe fallback pure objectField
-resolveField _ _ f = queryError ("Unexpected Selection value. Is the query normalized?: " <> show f)
+data NamedValueResolver m = NamedValueResolver AST.Name (m Result)
+
+
+
+-- Iterate through handlers (zipped together with their type
+-- definition) and execute handler if the name matches.
+--
+-- TODO: tuple return not great (but either is wrong, too because it discards errors.
+type ResolveFieldResult = ([QueryError], Maybe GValue.ObjectField)
```

Comment:
Re the TODO. Based on the comment on `Result` ("Result collects errors and values at the same time"), `Result` itself should have this ability. You can see this reflected in the structure of `Response` (which should probably _not_ be used for individual field results, since it's intended to represent the whole response). 

I wonder if it would be better just to go for that and remove this type alias. Doing so would at least move away from the anonymous product type and go to an explicit one.